File: src/main/java/spring/webmvc/infrastructure/aspect/RequestLockAspect.java
package spring.webmvc.infrastructure.aspect;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.springframework.stereotype.Component;

import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import spring.webmvc.application.service.RequestLockService;
import spring.webmvc.infrastructure.util.SecurityContextUtil;

@Aspect
@Component
@RequiredArgsConstructor
public class RequestLockAspect {

	private final HttpServletRequest httpServletRequest;
	private final RequestLockService requestLockService;

	@Pointcut("@annotation(spring.webmvc.infrastructure.aspect.RequestLock)")
	public void preventDuplicateRequestAnnotation() {
	}

	@Pointcut("within(@org.springframework.web.bind.annotation.RestController *)")
	public void withinRestController() {
	}

	@Pointcut("within(@org.springframework.security.access.prepost.PreAuthorize *)")
	public void withinPreAuthorize() {
	}

	@Around("preventDuplicateRequestAnnotation() && withinRestController() && withinPreAuthorize()")
	public Object preventDuplicateRequest(ProceedingJoinPoint joinPoint) throws Throwable {
		String uri = httpServletRequest.getRequestURI();
		String method = httpServletRequest.getMethod();
		Long memberId = SecurityContextUtil.getMemberId();

		requestLockService.validate(memberId, method, uri);

		return joinPoint.proceed();
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/infrastructure/aspect/RequestLockAspect.kt
package spring.web.kotlin.infrastructure.aspect

import jakarta.servlet.http.HttpServletRequest
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.springframework.stereotype.Component
import spring.web.kotlin.application.service.RequestLockService
import spring.web.kotlin.infrastructure.util.SecurityContextUtil

@Aspect
@Component
class RequestLockAspect(
    private val httpServletRequest: HttpServletRequest,
    private val requestLockService: RequestLockService
) {

    @Pointcut("@annotation(spring.web.kotlin.infrastructure.aspect.RequestLock)")
    fun preventDuplicateRequestAnnotation() {
    }

    @Pointcut("within(@org.springframework.web.bind.annotation.RestController *)")
    fun withinRestController() {
    }

    @Pointcut("within(@org.springframework.security.access.prepost.PreAuthorize *)")
    fun withinPreAuthorize() {
    }

    @Around("preventDuplicateRequestAnnotation() && withinRestController() && withinPreAuthorize()")
    fun preventDuplicateRequest(joinPoint: ProceedingJoinPoint): Any {
        val uri = httpServletRequest.requestURI
        val method = httpServletRequest.method
        val memberId = SecurityContextUtil.getMemberId()

        requestLockService.validate(memberId, method, uri)

        return joinPoint.proceed()
    }
}